src/libraries/System.Net.Http/src/System/Net/Http/MetricsHandler.cs

```diff
@@ -0,0 +1,235 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Net.Http.Metrics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http;
+
+internal sealed class MetricsHandler : HttpMessageHandlerStage, IHttpMetricsLogger
+{
+    private readonly HttpMessageHandler _innerHandler;
+    private readonly UpDownCounter<long> _currentRequests;
+    private readonly Counter<long> _failedRequests;
+    private readonly Histogram<double> _requestsDuration;
+
+    [ThreadStatic]
+    private static HttpMetricsEnrichmentContext? s_cachedEnrichmentContext;
+
+    public MetricsHandler(HttpMessageHandler innerHandler, Meter meter)
+    {
+        _innerHandler = innerHandler;
+
+        // Meter has a cache for
+        _currentRequests = meter.CreateUpDownCounter<long>(
+            "http-client-current-requests",
+            description: "Number of outbound HTTP requests that are currently active on the client.");
+        _failedRequests = meter.CreateCounter<long>(
+            "http-client-failed-requests",
+            description: "Number of outbound HTTP requests that have failed.");
+        _requestsDuration = meter.CreateHistogram<double>(
+            "http-client-request-duration",
+            unit: "s",
+            description: "The duration of outbound HTTP requests.");
+    }
+
+    public static Meter DefaultMeter { get; } = new SharedMeter();
+
+    internal override ValueTask<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+    {
+        if (_currentRequests.Enabled || _failedRequests.Enabled || _requestsDuration.Enabled)
+        {
+            ArgumentNullException.ThrowIfNull(request);
+            return SendAsyncWithMetrics(request, async, cancellationToken);
+        }
+        else
+        {
+            return async ?
+                new ValueTask<HttpResponseMessage>(_innerHandler.SendAsync(request, cancellationToken)) :
+                new ValueTask<HttpResponseMessage>(_innerHandler.Send(request, cancellationToken));
```

Comment:
```suggestion
            return new ValueTask<HttpResponseMessage>(async ?
                _innerHandler.SendAsync(request, cancellationToken) :
                _innerHandler.Send(request, cancellationToken));
```

Reply:
The two branches are calling two different overloads of the `ValueTask` ctr. (`ValueTask<Task<TResult>>` vs `ValueTask(TResult)`.

Reply:
Yup